File: cli/src/lib/fuzzy.ts
/**
 * MRSF Fuzzy Matching Engine
 *
 * Provides exact, normalized, token-level LCS, and character-level
 * Levenshtein matching for re-anchoring selected_text.
 */

import { distance as levenshtein } from "fastest-levenshtein";
import type { FuzzyCandidate } from "./types.js";

export const MAX_FUZZY_CANDIDATE_LINES = 64;

export interface FuzzySearchIndex {
  lines: string[];
  tokenPostings: Map<string, number[]>;
  trigramPostings: Map<string, number[]>;
}

export function createFuzzySearchIndex(lines: string[]): FuzzySearchIndex {
  const tokenPostings = new Map<string, number[]>();
  const trigramPostings = new Map<string, number[]>();
  for (let line = 1; line < lines.length; line += 1) {
    addPostingSignals(tokenPostings, lexicalTokens(lines[line]), line);
    addPostingSignals(trigramPostings, characterTrigrams(lines[line]), line);
  }
  return { lines, tokenPostings, trigramPostings };
}

// ---------------------------------------------------------------------------
// Exact matching
// ---------------------------------------------------------------------------

/**
 * Find all exact occurrences of `needle` in lines (1-based array).
 */
export function exactMatch(
  lines: string[],
  needle: string,
): FuzzyCandidate[] {
  if (!needle) return [];

  const results: FuzzyCandidate[] = [];
  const needleLines = needle.split("\n");
  const needleLineCount = needleLines.length;

  // Slide a window across the document
  for (let startLine = 1; startLine <= lines.length - needleLineCount; startLine++) {
    // Build the text for this window
    const windowLines = lines.slice(startLine, startLine + needleLineCount);
    const windowText = windowLines.join("\n");

    // Check if the needle appears anywhere within this window (for single-line)
    if (needleLineCount === 1) {
      let col = 0;
      const line = windowLines[0];
      while (col < line.length) {
        const idx = line.indexOf(needle, col);
        if (idx === -1) break;
        results.push({
          text: needle,
          line: startLine,
          endLine: startLine,
          startColumn: idx,
          endColumn: idx + needle.length,
          score: 1.0,
        });
        col = idx + 1;
      }
    } else {
      // Multi-line: check if the window contains the exact needle
      const idx = windowText.indexOf(needle);
      if (idx !== -1) {
        // Calculate start column
        const beforeMatch = windowText.slice(0, idx);
        const linesBeforeEnd = beforeMatch.split("\n");
        const startCol = linesBeforeEnd[linesBeforeEnd.length - 1].length;

        // Calculate end column
        const afterMatch = needle.split("\n");
        const endCol = afterMatch[afterMatch.length - 1].length;
        if (startCol === 0 || linesBeforeEnd.length === 1) {
          results.push({
            text: needle,
            line: startLine + linesBeforeEnd.length - 1,
            endLine: startLine + linesBeforeEnd.length - 1 + afterMatch.length - 1,
            startColumn: startCol,
            endColumn: endCol,
            score: 1.0,
          });
        }
      }
    }
  }

  return results;
}

// ---------------------------------------------------------------------------
// Normalized matching (collapse whitespace)
// ---------------------------------------------------------------------------

function normalize(text: string): string {
  return text.replace(/\s+/g, " ").trim();
}

/**
 * Find matches after normalizing whitespace.
 */
export function normalizedMatch(
  lines: string[],
  needle: string,
): FuzzyCandidate[] {
  const normNeedle = normalize(needle);
  if (!normNeedle) return [];

  const results: FuzzyCandidate[] = [];

  // Try expanding windows of varying sizes
  const needleLineEstimate = needle.split("\n").length;
  const minWindow = Math.max(1, needleLineEstimate - 1);
  const maxWindow = Math.min(lines.length - 1, needleLineEstimate + 2);

  for (let winSize = minWindow; winSize <= maxWindow; winSize++) {
    for (let startLine = 1; startLine + winSize - 1 < lines.length; startLine++) {
      const windowLines = lines.slice(startLine, startLine + winSize);
      const windowText = windowLines.join("\n");
      const normWindow = normalize(windowText);

      if (normWindow.includes(normNeedle)) {
        results.push({
          text: windowText,
          line: startLine,
          endLine: startLine + winSize - 1,
          startColumn: 0,
          endColumn: windowLines[windowLines.length - 1].length,
          score: 0.95,
        });
      }
    }
  }

  return deduplicateCandidates(results);
}

// ---------------------------------------------------------------------------
// Token-level LCS
// ---------------------------------------------------------------------------

function tokenize(text: string): string[] {
  return text.split(/\s+/).filter((t) => t.length > 0);
}

/**
 * Longest Common Subsequence length of two token arrays.
 */
function lcsLength(a: string[], b: string[]): number {
  const m = a.length;
  const n = b.length;
  // Optimize: use two rows instead of full matrix
  let prev = new Array<number>(n + 1).fill(0);
  let curr = new Array<number>(n + 1).fill(0);

  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      if (a[i - 1] === b[j - 1]) {
        curr[j] = prev[j - 1] + 1;
      } else {
        curr[j] = Math.max(prev[j], curr[j - 1]);
      }
    }
    [prev, curr] = [curr, prev];
    curr.fill(0);
  }

  return prev[n];
}

/**
 * Score two texts using token-level LCS.
 * Returns 0.0–1.0.
 */
export function tokenLcsScore(a: string, b: string): number {
  const tokA = tokenize(a);
  const tokB = tokenize(b);
  if (tokA.length === 0 && tokB.length === 0) return 1.0;
  if (tokA.length === 0 || tokB.length === 0) return 0.0;
  const lcs = lcsLength(tokA, tokB);
  return lcs / Math.max(tokA.length, tokB.length);
}

// ---------------------------------------------------------------------------
// Character-level Levenshtein score
// ---------------------------------------------------------------------------

/**
 * Normalized Levenshtein similarity 0.0–1.0.
 */
export function levenshteinScore(a: string, b: string): number {
  if (a.length === 0 && b.length === 0) return 1.0;
  const maxLen = Math.max(a.length, b.length);
  if (maxLen === 0) return 1.0;
  const dist = levenshtein(a, b);
  return 1 - dist / maxLen;
}

// ---------------------------------------------------------------------------
// Combined fuzzy scoring
// ---------------------------------------------------------------------------

/**
 * Compute a combined similarity score between two text fragments.
 * Blends token LCS (structural) and Levenshtein (character-level).
 */
export function combinedScore(needle: string, candidate: string): number {
  const tScore = tokenLcsScore(needle, candidate);

  // Full Levenshtein is expensive for long texts; only use for short ones
  let lScore: number;
  if (needle.length < 500 && candidate.length < 500) {
    lScore = levenshteinScore(needle, candidate);
  } else {
    lScore = tScore; // fall back to token score only
  }

  // Weight: 60% token LCS, 40% Levenshtein
  return tScore * 0.6 + lScore * 0.4;
}

// ---------------------------------------------------------------------------
// Fuzzy search across document
// ---------------------------------------------------------------------------

/**
 * Search the document for fuzzy matches of `needle`.
 *
 * @param lines     1-based line array (index 0 unused).
 * @param needle    The original selected_text.
 * @param threshold Minimum score to include (0.0–1.0).
 * @param hintLine  Optional original line number for proximity scoring.
 */
export function fuzzySearch(
  lines: string[],
  needle: string,
  threshold: number = 0.6,
  hintLine?: number,
  index?: FuzzySearchIndex,
): FuzzyCandidate[] {
  return fuzzySearchThresholds(
    lines,
    needle,
    [threshold],
    hintLine,
    index,
  ).get(threshold) ?? [];
}

/**
 * Compute fuzzy candidates once and partition them by their unadjusted
 * similarity thresholds. Proximity remains a ranking bonus and does not make
 * a candidate eligible for a threshold it did not originally satisfy.
 */
export function fuzzySearchThresholds(
  lines: string[],
  needle: string,
  thresholds: number[],
  hintLine?: number,
  index: FuzzySearchIndex = createFuzzySearchIndex(lines),
): Map<number, FuzzyCandidate[]> {
  const uniqueThresholds = [...new Set(thresholds)];
  const results = new Map<number, FuzzyCandidate[]>();
  if (uniqueThresholds.length === 0) return results;
  if (!needle) {
    for (const threshold of uniqueThresholds) results.set(threshold, []);
    return results;
  }

  const needleLines = needle.split("\n");
  const needleLineCount = needleLines.length;
  const candidates: FuzzyCandidate[] = [];
  const minimumThreshold = Math.min(...uniqueThresholds);
  const candidateLines = retrieveCandidateLines(index, needle, hintLine);

  // Window sizes: ±30% of original line count, minimum 1
  const minWindow = Math.max(1, Math.floor(needleLineCount * 0.7));
  const maxWindow = Math.min(
    lines.length - 1,
    Math.ceil(needleLineCount * 1.3) + 1,
  );

  for (let winSize = minWindow; winSize <= maxWindow; winSize++) {
    const startLines = new Set<number>();
    for (const candidateLine of candidateLines) {
      for (let offset = 0; offset < winSize; offset += 1) {
        const startLine = candidateLine - offset;
        if (startLine >= 1 && startLine + winSize - 1 < lines.length) {
          startLines.add(startLine);
        }
      }
    }
    for (const startLine of startLines) {
      const windowLines = lines.slice(startLine, startLine + winSize);
      const windowText = windowLines.join("\n");

      const score = combinedScore(needle, windowText);

      if (score >= minimumThreshold) {
        candidates.push({
          text: windowText,
          line: startLine,
          endLine: startLine + winSize - 1,
          startColumn: 0,
          endColumn: windowLines[windowLines.length - 1].length,
          score,
        });
      }
    }
  }

  // For single-line needles, also try substring matching within each line
  if (needleLineCount === 1 && needle.length < 200) {
    for (const lineNum of candidateLines) {
      const line = lines[lineNum];
      if (!line) continue;

      const winLen = needle.length;
      const minWinLen = Math.max(3, Math.floor(winLen * 0.7));
      const maxWinLen = Math.min(line.length, Math.ceil(winLen * 1.3));
      const lengths = evenlySpacedIntegers(minWinLen, maxWinLen, 7);
      for (const len of lengths) {
        for (let col = 0; col + len <= line.length; col++) {
          const sub = line.substring(col, col + len);
          const score = combinedScore(needle, sub);
          if (score >= minimumThreshold) {
            candidates.push({
              text: sub,
              line: lineNum,
              endLine: lineNum,
              startColumn: col,
              endColumn: col + len,
              score,
            });
          }
        }
      }
    }
  }

  const deduped = deduplicateCandidates(candidates);
  const scored = deduped.map((candidate) => ({
    baseScore: candidate.score,
    candidate: applyProximityBonus(candidate, hintLine),
  }));

  for (const threshold of uniqueThresholds) {
    results.set(
      threshold,
      scored
        .filter((item) => item.baseScore >= threshold)
        .map((item) => item.candidate)
        .sort((left, right) => right.score - left.score),
    );
  }

  return results;
}

function retrieveCandidateLines(
  index: FuzzySearchIndex,
  needle: string,
  hintLine?: number,
): number[] {
  const votes = new Map<number, number>();
  const lineCount = Math.max(0, index.lines.length - 1);
  const signals = [
    ...postingSignals(index.tokenPostings, lexicalTokens(needle), 2),
    ...postingSignals(index.trigramPostings, characterTrigrams(needle), 1),
  ]
    .sort((left, right) => left.postings.length - right.postings.length)
    .slice(0, 16);

  for (const signal of signals) {
    const rarity = Math.log1p(lineCount / signal.postings.length);
    for (const line of signal.postings) {
      votes.set(line, (votes.get(line) ?? 0) + signal.weight * rarity);
    }
  }

  if (hintLine != null) {
    for (let offset = -4; offset <= 4; offset += 1) {
      const line = hintLine + offset;
      if (line >= 1 && line <= lineCount) {
        votes.set(line, (votes.get(line) ?? 0) + 0.25);
      }
    }
  }

  return [...votes.entries()]
    .sort((left, right) =>
      right[1] - left[1]
      || distanceFromHint(left[0], hintLine) - distanceFromHint(right[0], hintLine)
      || left[0] - right[0]
    )
    .slice(0, MAX_FUZZY_CANDIDATE_LINES)
    .map(([line]) => line);
}

function postingSignals(
  postings: Map<string, number[]>,
  values: string[],
  weight: number,
): Array<{ postings: number[]; weight: number }> {
  return [...new Set(values)]
    .map((value) => ({ postings: postings.get(value) ?? [], weight }))
    .filter((signal) => signal.postings.length > 0);
}

function addPostingSignals(
  postings: Map<string, number[]>,
  values: string[],
  line: number,
): void {
  for (const value of new Set(values)) {
    const lines = postings.get(value);
    if (lines) {
      lines.push(line);
    } else {
      postings.set(value, [line]);
    }
  }
}

function lexicalTokens(text: string): string[] {
  return text.toLowerCase().match(/[\p{L}\p{N}_-]+/gu) ?? [];
}

function characterTrigrams(text: string): string[] {
  const normalized = text.toLowerCase().replace(/\s+/g, " ").trim();
  const characters = [...normalized];
  if (characters.length < 3) return normalized ? [normalized] : [];
  const trigrams: string[] = [];
  for (let index = 0; index <= characters.length - 3; index += 1) {
    trigrams.push(characters.slice(index, index + 3).join(""));
  }
  return trigrams;
}

function evenlySpacedIntegers(
  minimum: number,
  maximum: number,
  count: number,
): number[] {
  if (maximum <= minimum) return [minimum];
  const values = new Set<number>();
  for (let index = 0; index < count; index += 1) {
    values.add(Math.round(minimum + (maximum - minimum) * index / (count - 1)));
  }
  return [...values];
}

function distanceFromHint(line: number, hintLine?: number): number {
  return hintLine == null ? 0 : Math.abs(line - hintLine);
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function deduplicateCandidates(
  candidates: FuzzyCandidate[],
): FuzzyCandidate[] {
  const seen = new Map<string, FuzzyCandidate>();
  for (const c of candidates) {
    const key = `${c.line}:${c.startColumn}:${c.endLine}:${c.endColumn}`;
    const existing = seen.get(key);
    if (!existing || c.score > existing.score) {
      seen.set(key, c);
    }
  }
  return Array.from(seen.values());
}

function applyProximityBonus(
  candidate: FuzzyCandidate,
  hintLine?: number,
): FuzzyCandidate {
  if (hintLine == null) return candidate;

  const distance = Math.abs(candidate.line - hintLine);
  const proximityBonus = 0.1 * Math.max(0, 1 - distance / 50);
  return {
    ...candidate,
    score: Math.min(1.0, candidate.score + proximityBonus),
  };
}

File: python/src/mrsf/fuzzy.py
"""MRSF Fuzzy Matching Engine.

Provides exact, normalized, token-level LCS, and character-level
Levenshtein matching for re-anchoring selected_text.
"""

from __future__ import annotations

import math
import re
from dataclasses import dataclass

from rapidfuzz.distance import Levenshtein

from .types import FuzzyCandidate

MAX_FUZZY_CANDIDATE_LINES = 64


@dataclass
class FuzzySearchIndex:
    lines: list[str]
    token_postings: dict[str, list[int]]
    trigram_postings: dict[str, list[int]]


def create_fuzzy_search_index(lines: list[str]) -> FuzzySearchIndex:
    tokens: dict[str, list[int]] = {}
    trigrams: dict[str, list[int]] = {}
    for line in range(1, len(lines)):
        _add_postings(tokens, _lexical_tokens(lines[line]), line)
        _add_postings(trigrams, _character_trigrams(lines[line]), line)
    return FuzzySearchIndex(lines, tokens, trigrams)

# ---------------------------------------------------------------------------
# Exact matching
# ---------------------------------------------------------------------------


def exact_match(lines: list[str], needle: str) -> list[FuzzyCandidate]:
    """Find all exact occurrences of `needle` in lines (1-based array)."""
    if not needle:
        return []

    results: list[FuzzyCandidate] = []
    needle_lines = needle.split("\n")
    needle_line_count = len(needle_lines)

    for start_line in range(1, len(lines) - needle_line_count + 1):
        window_lines = lines[start_line : start_line + needle_line_count]
        window_text = "\n".join(window_lines)

        if needle_line_count == 1:
            line = window_lines[0]
            col = 0
            while col < len(line):
                idx = line.find(needle, col)
                if idx == -1:
                    break
                results.append(
                    FuzzyCandidate(
                        text=needle,
                        line=start_line,
                        end_line=start_line,
                        start_column=idx,
                        end_column=idx + len(needle),
                        score=1.0,
                    )
                )
                col = idx + 1
        else:
            idx = window_text.find(needle)
            if idx != -1:
                before_match = window_text[:idx]
                lines_before_end = before_match.split("\n")
                start_col = len(lines_before_end[-1])

                after_match = needle.split("\n")
                end_col = len(after_match[-1])
                if start_col == 0 or len(lines_before_end) == 1:
                    results.append(
                        FuzzyCandidate(
                            text=needle,
                            line=start_line + len(lines_before_end) - 1,
                            end_line=start_line + len(lines_before_end) - 1 + len(after_match) - 1,
                            start_column=start_col,
                            end_column=end_col,
                            score=1.0,
                        )
                    )

    return results


# ---------------------------------------------------------------------------
# Normalized matching
# ---------------------------------------------------------------------------


def _normalize(text: str) -> str:
    return re.sub(r"\s+", " ", text).strip()


def normalized_match(lines: list[str], needle: str) -> list[FuzzyCandidate]:
    """Find matches after normalizing whitespace."""
    norm_needle = _normalize(needle)
    if not norm_needle:
        return []

    results: list[FuzzyCandidate] = []
    needle_line_estimate = len(needle.split("\n"))
    min_window = max(1, needle_line_estimate - 1)
    max_window = min(len(lines) - 1, needle_line_estimate + 2)

    for win_size in range(min_window, max_window + 1):
        for start_line in range(1, len(lines) - win_size + 1):
            window_lines = lines[start_line : start_line + win_size]
            window_text = "\n".join(window_lines)
            norm_window = _normalize(window_text)

            if norm_needle in norm_window:
                results.append(
                    FuzzyCandidate(
                        text=window_text,
                        line=start_line,
                        end_line=start_line + win_size - 1,
                        start_column=0,
                        end_column=len(window_lines[-1]),
                        score=0.95,
                    )
                )

    return _deduplicate_candidates(results)


# ---------------------------------------------------------------------------
# Token-level LCS
# ---------------------------------------------------------------------------


def _tokenize(text: str) -> list[str]:
    return text.split()


def _lcs_length(a: list[str], b: list[str]) -> int:
    m, n = len(a), len(b)
    prev = [0] * (n + 1)
    curr = [0] * (n + 1)

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if a[i - 1] == b[j - 1]:
                curr[j] = prev[j - 1] + 1
            else:
                curr[j] = max(prev[j], curr[j - 1])
        prev, curr = curr, [0] * (n + 1)

    return prev[n]


def token_lcs_score(a: str, b: str) -> float:
    """Score two texts using token-level LCS. Returns 0.0–1.0."""
    tok_a = _tokenize(a)
    tok_b = _tokenize(b)
    if not tok_a and not tok_b:
        return 1.0
    if not tok_a or not tok_b:
        return 0.0
    lcs = _lcs_length(tok_a, tok_b)
    return lcs / max(len(tok_a), len(tok_b))


# ---------------------------------------------------------------------------
# Character-level Levenshtein score
# ---------------------------------------------------------------------------


def levenshtein_score(a: str, b: str) -> float:
    """Normalized Levenshtein similarity 0.0–1.0."""
    if not a and not b:
        return 1.0
    max_len = max(len(a), len(b))
    if max_len == 0:
        return 1.0
    dist = Levenshtein.distance(a, b)
    return 1.0 - dist / max_len


# ---------------------------------------------------------------------------
# Combined fuzzy scoring
# ---------------------------------------------------------------------------


def combined_score(needle: str, candidate: str) -> float:
    """Compute a combined similarity score (60% token LCS, 40% Levenshtein)."""
    t_score = token_lcs_score(needle, candidate)

    if len(needle) < 500 and len(candidate) < 500:
        l_score = levenshtein_score(needle, candidate)
    else:
        l_score = t_score

    return t_score * 0.6 + l_score * 0.4


# ---------------------------------------------------------------------------
# Fuzzy search across document
# ---------------------------------------------------------------------------


def fuzzy_search(
    lines: list[str],
    needle: str,
    threshold: float = 0.6,
    hint_line: int | None = None,
    index: FuzzySearchIndex | None = None,
) -> list[FuzzyCandidate]:
    """Search the document for fuzzy matches of needle."""
    return fuzzy_search_thresholds(
        lines, needle, [threshold], hint_line, index
    ).get(threshold, [])


def fuzzy_search_thresholds(
    lines: list[str],
    needle: str,
    thresholds: list[float],
    hint_line: int | None = None,
    index: FuzzySearchIndex | None = None,
) -> dict[float, list[FuzzyCandidate]]:
    """Compute bounded fuzzy candidates once and partition by base threshold."""
    unique_thresholds = list(dict.fromkeys(thresholds))
    results: dict[float, list[FuzzyCandidate]] = {}
    if not unique_thresholds:
        return results
    if not needle:
        return {threshold: [] for threshold in unique_thresholds}
    if index is None:
        index = create_fuzzy_search_index(lines)

    needle_lines = needle.split("\n")
    needle_line_count = len(needle_lines)
    candidates: list[FuzzyCandidate] = []
    minimum_threshold = min(unique_thresholds)
    candidate_lines = _retrieve_candidate_lines(index, needle, hint_line)

    min_window = max(1, int(needle_line_count * 0.7))
    max_window = min(len(lines) - 1, int(needle_line_count * 1.3) + 1)

    for win_size in range(min_window, max_window + 1):
        start_lines = {
            candidate_line - offset
            for candidate_line in candidate_lines
            for offset in range(win_size)
            if candidate_line - offset >= 1
            and candidate_line - offset + win_size - 1 < len(lines)
        }
        for start_line in start_lines:
            window_lines = lines[start_line : start_line + win_size]
            window_text = "\n".join(window_lines)

            score = combined_score(needle, window_text)

            if score >= minimum_threshold:
                candidates.append(
                    FuzzyCandidate(
                        text=window_text,
                        line=start_line,
                        end_line=start_line + win_size - 1,
                        start_column=0,
                        end_column=len(window_lines[-1]),
                        score=score,
                    )
                )

    # Single-line substring matching
    if needle_line_count == 1 and len(needle) < 200:
        for line_num in candidate_lines:
            line = lines[line_num]
            if not line:
                continue

            win_len = len(needle)
            min_win_len = max(3, int(win_len * 0.7))
            max_win_len = min(len(line), int(win_len * 1.3))

            for length in _evenly_spaced(min_win_len, max_win_len, 7):
                for col in range(0, len(line) - length + 1):
                    sub = line[col : col + length]
                    score = combined_score(needle, sub)
                    if score >= minimum_threshold:
                        candidates.append(
                            FuzzyCandidate(
                                text=sub,
                                line=line_num,
                                end_line=line_num,
                                start_column=col,
                                end_column=col + length,
                                score=score,
                            )
                        )

    scored: list[tuple[float, FuzzyCandidate]] = []
    for candidate in _deduplicate_candidates(candidates):
        base_score = candidate.score
        if hint_line is not None:
            candidate = FuzzyCandidate(
                text=candidate.text,
                line=candidate.line,
                end_line=candidate.end_line,
                start_column=candidate.start_column,
                end_column=candidate.end_column,
                score=min(
                    1.0,
                    candidate.score
                    + 0.1 * max(0, 1 - abs(candidate.line - hint_line) / 50),
                ),
            )
        scored.append((base_score, candidate))
    for threshold in unique_thresholds:
        results[threshold] = sorted(
            (candidate for base, candidate in scored if base >= threshold),
            key=lambda candidate: candidate.score,
            reverse=True,
        )
    return results


def _retrieve_candidate_lines(
    index: FuzzySearchIndex,
    needle: str,
    hint_line: int | None,
) -> list[int]:
    votes: dict[int, float] = {}
    line_count = max(0, len(index.lines) - 1)
    signals = [
        *(_posting_signals(index.token_postings, _lexical_tokens(needle), 2.0)),
        *(_posting_signals(index.trigram_postings, _character_trigrams(needle), 1.0)),
    ]
    signals.sort(key=lambda signal: len(signal[0]))
    for postings, weight in signals[:16]:
        rarity = math.log1p(line_count / len(postings))
        for line in postings:
            votes[line] = votes.get(line, 0.0) + weight * rarity
    if hint_line is not None:
        for offset in range(-4, 5):
            line = hint_line + offset
            if 1 <= line <= line_count:
                votes[line] = votes.get(line, 0.0) + 0.25
    return [
        line
        for line, _ in sorted(
            votes.items(),
            key=lambda item: (
                -item[1],
                abs(item[0] - hint_line) if hint_line is not None else 0,
                item[0],
            ),
        )[:MAX_FUZZY_CANDIDATE_LINES]
    ]


def _posting_signals(
    postings: dict[str, list[int]],
    values: list[str],
    weight: float,
) -> list[tuple[list[int], float]]:
    return [
        (postings[value], weight)
        for value in dict.fromkeys(values)
        if value in postings
    ]


def _add_postings(postings: dict[str, list[int]], values: list[str], line: int) -> None:
    for value in dict.fromkeys(values):
        postings.setdefault(value, []).append(line)


def _lexical_tokens(text: str) -> list[str]:
    return re.findall(r"[\w-]+", text.lower(), flags=re.UNICODE)


def _character_trigrams(text: str) -> list[str]:
    normalized = re.sub(r"\s+", " ", text.lower()).strip()
    if len(normalized) < 3:
        return [normalized] if normalized else []
    return [normalized[index : index + 3] for index in range(len(normalized) - 2)]


def _evenly_spaced(minimum: int, maximum: int, count: int) -> list[int]:
    if maximum <= minimum:
        return [minimum]
    return list(
        dict.fromkeys(
            round(minimum + (maximum - minimum) * index / (count - 1))
            for index in range(count)
        )
    )


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _deduplicate_candidates(candidates: list[FuzzyCandidate]) -> list[FuzzyCandidate]:
    seen: dict[str, FuzzyCandidate] = {}
    for c in candidates:
        key = f"{c.line}:{c.start_column}:{c.end_line}:{c.end_column}"
        existing = seen.get(key)
        if existing is None or c.score > existing.score:
            seen[key] = c
    return list(seen.values())
